weldx/util.py

```diff
@@ -1201,3 +1193,78 @@ def time_ref(self, value: pd.Timestamp):
                 self._obj.time.attrs["time_ref"] = value  # set new time_ref value
             else:
                 self._obj.time.attrs["time_ref"] = value
+
+
+class _Eq_compare_nested:
+    """Compares nested data structures like lists, sets, tuples, arrays, etc."""
+
+    compare_funcs = {
+        (np.ndarray, pint.Quantity): lambda x, y: np.all(x == y),
+        (xr.DataArray, xr.Dataset): lambda x, y: x.identical(y),
+    }
+
+    @staticmethod
+    def _compare(x, y):
+        if type(x) is not type(y):
+            return False
+        for types, func in _Eq_compare_nested.compare_funcs.items():
+            if isinstance(x, types):
+                return func(x, y)
+            return x == y
+
+    @staticmethod
+    def _visit(p, k, v, b):
+        """Traverses all elements in `compare_nested` argument a and b...
+
+        and tries to obtain the path `p` in `b` using boltons.iterutils.get_path.
+        The following cases can occur:
+        1. If the path does not exist in `b` a KeyError will be raised.
+        2. If the index `k` does not exist an IndexError is raised.
+        3. If the other path exists, a comparison will be made using `_compare`.
+           When the elements are not equal traversing `a` will be stopped
+           by raising a ValueError.
+        """
+        other_value = iterutils.get_path(b, p)[k]
+        if not _Eq_compare_nested._compare(v, other_value):
+            raise ValueError
+        return True
+
+    @staticmethod
+    def compare_nested(a, b):
+        """Deeply compares [nested] data structures combined of tuples, lists, dicts...
+
+        Also compares non-nested data-structures.
+        Arrays are compared using np.all and xr.DataArray.identical
+
+        Parameters
+        ----------
+        a :
+            a [nested] data structure to compare to `b`.
+        b :
+            a [nested] data structure to compare to `a`.
+
+        Returns
+        -------
+        bool :
+            True, if all elements (including dict keys) of a and b are equal.
+
+        Raises
+        ------
+        TypeError
+            When a or b is not a nested structure.
+
+        """
+        visit = functools.partial(_Eq_compare_nested._visit, b=b)
+
+        try:
+            iterutils.remap(a, visit=visit, reraise_visit=True)
+        # Key not found in b, values not equal, more elements in a than in b
+        except (KeyError, ValueError, IndexError):
+            return False
```

Comment:
```suggestion
            raise
```
I would prefer just to raise the error because otherwise this is really hard to debug if the Exception gets lost

Reply:
Key and Index errors are being raised if something is not accessible in the other data structure. IMO returning false here is correct. TypeErrors are re-raised.

Reply:
Agree with @marscher here. Exceptions that are raised during the comparison due to missing keys in one of the compared dicts shouldn't be propagated. The function should simply return not equal e.g. `False`.